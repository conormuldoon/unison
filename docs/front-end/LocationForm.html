<!DOCTYPE html>
<html lang="en">

<head>
  
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title> LocationForm</title>

  <script src="https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js"></script>
  <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
  <script src="./build/entry.js"></script>
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  <link href="https://fonts.googleapis.com/css?family=Roboto:100,400,700|Inconsolata,700" rel="stylesheet">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
  <link type="text/css" rel="stylesheet" href="https://jmblog.github.io/color-themes-for-google-code-prettify/themes/tomorrow-night.min.css">
  <link type="text/css" rel="stylesheet" href="styles/app.min.css">
  <link type="text/css" rel="stylesheet" href="styles/iframe.css">
  <link type="text/css" rel="stylesheet" href="">
  <script async defer src="https://buttons.github.io/buttons.js"></script>

  
</head>



<body class="layout small-header">
    <div id="stickyNavbarOverlay"></div>
    

<div class="top-nav">
    <div class="inner">
        <a id="hamburger" role="button" class="navbar-burger" aria-label="menu" aria-expanded="false">
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
        </a>
        <div class="logo">
            
             
                <a href="index.html">
                    <h1 class="navbar-item">Unison React components</h1>
                </a>
            
        </div>
        <div class="menu">
            
            <div class="navigation">
                <a
                    href="index.html"
                    class="link"
                >
                    API Documentation
                </a>
                
                
                
            </div>
        </div>
    </div>
</div>
    <div id="main">
        <div
            class="sidebar "
            id="sidebarNav"
        >
            
            <nav>
                
                    <h2><a href="index.html">Documentation</a></h2><div class="category"><h3>Components</h3><ul><li><a href="ARLocationComponent.html">ARLocationComponent</a></li><li><a href="ChartComponent.html">ChartComponent</a></li><li><a href="ChartPopup.html">ChartPopup</a></li><li><a href="LeafletMap.html">LeafletMap</a></li><li><a href="LocationForm.html">LocationForm</a></li><li><a href="RemoveComponent.html">RemoveComponent</a></li><li><a href="TabsComponent.html">TabsComponent</a></li><li><a href="Unison.html">Unison</a></li></ul><h3>Global</h3><ul><li><a href="global.html#DateSelector">DateSelector</a></li></ul></div>
                
            </nav>
        </div>
        <div class="core" id="main-content-wrapper">
            <div class="content">
                <header class="page-title">
                    <p>Components</p>
                    <h1>LocationForm</h1>
                </header>
                




<section>

<header>
    
        <h2>
            &ltLocationForm <i>display</i> <i>toggleDisplay</i> <i>hideDisplay</i> <i>obtainData</i> <i>featureProperties</i> <i>collectionModel</i> /&gt
        </h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        
            <div class='vertical-section'>
                <div class="members">
                    <div class="member">
                        <div class=name>
                            <span class="tag">Constructor</span>
                        </div>
                        


    
    <h4 class="name" id="LocationForm">
        <a class="href-link" href="#LocationForm">#</a>
        
        <span class="code-name">
            
                &ltLocationForm <i>display</i> <i>toggleDisplay</i> <i>hideDisplay</i> <i>obtainData</i> <i>featureProperties</i> <i>collectionModel</i> /&gt
            
        </span>
    </h4>
    

    
    
    <div class="description">
        <p>A component for displaying a form to add new locations to be tracked.</p>
    </div>
    







    
        <h5>PropTypes:</h5>
        
<div class="table-container">
  <table class="params table">
    <thead>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Required</th>
        <th>Description</th>
        <th class="last">Default</th>
    </tr>
    </thead>

    <tbody>
    
          <tr class="deep-level-0">
            <td class="name"><code>display</code></td>
            <td>
              bool
            </td>
            <td>Yes</td>
            <td>Determines whether the form for entering location name, longitude, and latitude values is displayed.</td>
            <td></td>
          </tr>
          
    
          <tr class="deep-level-0">
            <td class="name"><code>toggleDisplay</code></td>
            <td>
              func
            </td>
            <td>Yes</td>
            <td>Toggles whether the form is displayed. If the remove form is displayed, it will be hidden.</td>
            <td></td>
          </tr>
          
    
          <tr class="deep-level-0">
            <td class="name"><code>hideDisplay</code></td>
            <td>
              func
            </td>
            <td>Yes</td>
            <td>Called when a location has been successfully added to hide the form.</td>
            <td></td>
          </tr>
          
    
          <tr class="deep-level-0">
            <td class="name"><code>obtainData</code></td>
            <td>
              func
            </td>
            <td>Yes</td>
            <td>Called when a location has been successfully added to obtain the updated location list from the server and add the location to the map.</td>
            <td></td>
          </tr>
          
    
          <tr class="deep-level-0">
            <td class="name"><code>featureProperties</code></td>
            <td>
              object
            </td>
            <td>No</td>
            <td></td>
            <td></td>
          </tr>
          
    
          <tr class="deep-level-0">
            <td class="name"><code>collectionModel</code></td>
            <td>
              object
            </td>
            <td>No</td>
            <td></td>
            <td></td>
          </tr>
          
    
    </tbody>
  </table>
</div>
    







<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="LocationForm.js.html" class="button">View Source</a>
            <span>
                <a href="LocationForm.js.html">LocationForm.js</a>, <a href="LocationForm.js.html#line17">line 17</a>
            </span>
        </p>
    
</dl>






















                    </div>
                </div>
            </div>
        
    
    </div>
    
        
    
    

    

    

    

    

    

    

    

    

    
</article>

</section>




            </div>
            
            <footer class="footer">
                <div class="content has-text-centered">
                    <p>Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.6</a></p>
                    <p class="sidebar-created-by">
                        <a href="https://github.com/SoftwareBrothers/better-docs" target="_blank">BetterDocs theme</a> provided with <i class="fas fa-heart"></i> by 
                        <a href="http://softwarebrothers.co" target="_blank">SoftwareBrothers - JavaScript Development Agency</a>
                    </p>
                </div>
            </footer>
            
        </div>
        <div id="side-nav" class="side-nav">
        </div>
    </div>
<script src="scripts/app.min.js"></script>
<script>PR.prettyPrint();</script>
<script src="scripts/linenumber.js"> </script>

</body>
</html>