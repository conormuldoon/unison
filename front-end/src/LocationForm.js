
import React, { useState } from 'react';
import {API} from './Constant';
import './App.css';
import {locationPostObject} from './Util';

import PropTypes from 'prop-types';


/**
 * A component for displaying a form to add new locations to be tracked.
 *  
 */
export default function LocationForm(props) {

  

  const [location,setLocation]=useState('');
  const [lon,setLon]=useState('');
  const [lat,setLat]=useState('');
  const [uname,setUName]=useState('');
  const [pword,setPWord]=useState('');


  function handleLoc(event){
    setLocation(event.target.value);

  }

  function handleLon(event){
    setLon(event.target.value);
  }

  function handleLat(event){
    setLat(event.target.value);

  }

  function handleUN(event){
    setUName(event.target.value);
  }

  function handlePassw(event){
    setPWord(event.target.value);
  }

  function handleSubmit(event){

    postData();

    event.preventDefault();

   }

   async function postData(){

    const response=await fetch(API+'/addLocation',
      locationPostObject(location,uname,pword,lon,lat));



    if(response.ok){
      const resVal=await response.json();

      if(resVal===0){
        alert('Location already exists');
      }else if(resVal===1){
        alert(location+' added');
        updateDisplay();

      }else if(resVal===2){
        alert('Incorrect user name or password');
      }else{
        alert(location + ' added, but did not obtain weather data.')
        updateDisplay();
      }
    }else{
      alert('Problem connecting');
    }

  }



  function updateDisplay(){
    setLocation('');
    setLon('');
    setLat('');
    setUName('');
    setPWord('');
    props.hideDisplay();
    props.obtainData();
  }


    let text='Add Location';
    if(props.display){
      text='Hide';
    }



    return ( <div id='pLeft'>

          <button data-testid="lf-button" onClick={props.toggleDisplay}>{text}</button>

          {props.display&&<form  id='locForm' onSubmit={handleSubmit}>
              <label>
                Location name: <input type="text" name="name" value={location} onChange={handleLoc} required  />
              </label>
              <br/>
              <br/>
              <label>
                Longitude: <input type="number" step="any" name="lon" value={lon} onChange={handleLon} required  />
              </label>
              <br/>
              <br/>
              <label>
                Latitude: <input type="number" step="any" name="lat"  value={lat} onChange={handleLat} required  />
              </label>
              <br/>
              <br/>
              <label>
                Username: <input type="text" name="un" onChange={handleUN}  value={uname}  required />
              </label>
              <br/>
              <br/>
              <label>
                Password: <input type="password" name="passwd" onChange={handlePassw}  value={pword} required />
              </label>
              <br/>
              <br/>
              <div>
                  <input type="submit" value="Submit" />
              </div>
            </form>}
        </div>


    );

}

LocationForm.propTypes ={
  /** Determines whether the form for entering location name, longitude, and latitude values is displayed. */
  display: PropTypes.bool,

  /** Toggles whether the form is displayed. If the remove form is displayed, it will be hidden. */
  toggleDisplay: PropTypes.func,

  /** Called when a location has been successfully added to hide the form. */
  hideDisplay: PropTypes.func,

  /** Called when a location has been successfully added to obtain the updated location list from the server and add the location to the map. */
  obtainData: PropTypes.func,
}
