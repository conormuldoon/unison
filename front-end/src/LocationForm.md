LocationForm example:

```js
import {useState} from 'react';
const [display,setDisplay] = useState(false);
function toggleDisplay(){
    if(display)setDisplay(false);
    else setDisplay(true);
}

<div style={{paddingLeft:200}}>
    <LocationForm display={display} toggleDisplay ={toggleDisplay} />
</div>
```