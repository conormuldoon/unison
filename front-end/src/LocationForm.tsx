import PropTypes from 'prop-types';
import React, { useState } from 'react';
import './App.css';

import { problemConnecting, csrfToken } from './Util';
import HttpStatus from 'http-status-codes';
import { locationPutObject } from './Util';

import parser from 'uri-template';

export function createLocationFactory(obtainData: () => void, selfRef: string, containsRef: string) {

  return function locationFactory(toggleDisplayAdd: () => void, displayAdd: boolean,
    hideAdd: () => void): JSX.Element {

    return <LocationForm obtainData={obtainData} toggleDisplay={toggleDisplayAdd}
      display={displayAdd} hideDisplay={hideAdd} selfRef={selfRef} containsRef={containsRef} />;
  }
}

export interface LocationProps {

  /**
   * A flag that determines whether the add location popup is displayed.
   */
  display: boolean;

  /**
   * A callback invoked to hide the add locatoin popup.
   */
  hideDisplay: () => void;

  /**
   * Toggles whether the add location popup is displayed.
   */
  toggleDisplay: () => void;

  /**
   * A callback for obtaining data from the server that is invoked once a location has been added.
   */
  obtainData: () => void;

  /**
   * A URI for new location PUT requests.
   */
  selfRef: string;

  /**
   * A URI for determining whether the location collection contains a location of a given name.
   */
  containsRef: string;
}

/**
 * A component for displaying a form to add new locations to be tracked.
 * 
 * @component 
 */
function LocationForm({ display, hideDisplay, toggleDisplay, obtainData, selfRef, containsRef }:
  LocationProps): JSX.Element {



  const [location, setLocation] = useState('');
  const [lon, setLon] = useState('');
  const [lat, setLat] = useState('');

  function handleLoc(event: React.ChangeEvent<HTMLInputElement>) {
    setLocation(event.target.value);

  }

  function handleLon(event: React.ChangeEvent<HTMLInputElement>) {
    setLon(event.target.value);
  }

  function handleLat(event: React.ChangeEvent<HTMLInputElement>) {
    setLat(event.target.value);

  }

  function updateDisplay() {
    setLocation('');
    setLon('');
    setLat('');
    hideDisplay();
    obtainData();
  }

  async function putData(selfRef: string) {

    const response = await fetch(selfRef,
      locationPutObject(location, lon, lat));



    if (response.ok) {

      const harvestResponse = await fetch(selfRef + '/' + location, {
        method: 'POST',
        headers: new Headers({
          'X-XSRF-TOKEN': csrfToken()
        }),
      });

      const message = (response.status === HttpStatus.OK) ? 'updated' : 'added';

      if (harvestResponse.ok) {
        alert(location + ' was ' + message + '.');
        updateDisplay();
      } else if (harvestResponse.status === HttpStatus.BAD_GATEWAY) {
        alert(location + ' was ' + message + ', but did did not recieve the weather data. The ' +
          lon + " and " + lat + " longitude and latitude coordinates may not be covered by the model.");
        updateDisplay();
      }
      else {
        alert(location + ' was ' + message + ', but Unison did not obtain the weather data.')
        updateDisplay();
      }
    } else if (response.status === HttpStatus.UNAUTHORIZED) {
      alert('Incorrect user name or password');
    } else if (response.status === HttpStatus.FORBIDDEN) {
      alert('You do not have permission to update ' + location +
        '. A location of the same name may have been added by another user.');
    } else {
      problemConnecting();
    }

  }

  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();

    const template = parser.parse(containsRef);
    const uri = template.expand({ name: location });

    const response = await fetch(uri);

    if (response.ok) {

      const contains = await response.json();
      if (contains.value) {

        if (window.confirm('A location named ' + location + " was previously added. Would you like to update it?")) {
          putData(selfRef);
        }

      } else {
        putData(selfRef);
      }

    } else {
      problemConnecting();
    }


  }


  let text = 'Add Location';
  if (display) {
    text = 'Hide';
  }


  return (<div id='pLeft'>

    <button data-testid="lf-button" onClick={toggleDisplay}>{text}</button>

    {display && <form id='locForm' onSubmit={handleSubmit}>
      <label>
        Location name: <input type="text" name="name" value={location} onChange={handleLoc} required />
      </label>
      <br />
      <br />
      <label>
        Longitude: <input type="number" step="any" name="lon" value={lon} onChange={handleLon} required />
      </label>
      <br />
      <br />
      <label>
        Latitude: <input type="number" step="any" name="lat" value={lat} onChange={handleLat} required />
      </label>
      <br />
      <br />
      <div>
        <input type="submit" value="Submit" />
      </div>
    </form>}
  </div>


  );

}

LocationForm.propTypes = {
  /** Determines whether the form for entering location name, longitude, and latitude values is displayed. */
  display: PropTypes.bool.isRequired,

  /** Toggles whether the form is displayed. If the remove form is displayed, it will be hidden. */
  toggleDisplay: PropTypes.func.isRequired,

  /** Called when a location has been successfully added to hide the form. */
  hideDisplay: PropTypes.func.isRequired,

  /** Called when a location has been successfully added to obtain the updated location list from the server and add the location to the map. */
  obtainData: PropTypes.func.isRequired,

  featureProperties: PropTypes.object,

  selfRef: PropTypes.string,

  containsRef: PropTypes.string,

}

export default LocationForm;
