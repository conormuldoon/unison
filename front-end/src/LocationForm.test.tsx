
import { enableFetchMocks } from 'jest-fetch-mock';
enableFetchMocks();
import fetchMock from 'jest-fetch-mock';


import { fireEvent, render, screen, act } from "@testing-library/react";
import LocationForm from './LocationForm';
import HttpStatus from 'http-status-codes';


it('renders without crashing', async () => {

  render(<LocationForm obtainData={() => { }} hideDisplay={() => { }} toggleDisplay={() => { }}
    display={true} selfRef="" containsRef="" />);

});


it('mathes snapshot', () => {


  const { container } = render(<LocationForm obtainData={() => { }} hideDisplay={() => { }}
    toggleDisplay={() => { }} display={true} selfRef="" containsRef="" />);

  expect(container).toMatchSnapshot();

});

async function changeValue(labelText: string, value: string) {

  const item = await screen.findByLabelText(labelText);
  await act(() => fireEvent.change(item, { target: { value: value } }));
}

it('sends the data in the form when submit is clicked', async () => {


  fetchMock.mockOnce(JSON.stringify({ value: false }));

  const hideDisplay = jest.fn();

  const location = 'UCD';

  const lon = '-6.223682';
  const lat = '53.308441';

  const obtainData = () => {


    expect(hideDisplay).toHaveBeenCalledTimes(1);

  };


  const component = <LocationForm display={true} containsRef={"http://localhost:8080/locationCollection/contains{?name}"} selfRef={"http://localhost:8080/locationCollection"} obtainData={obtainData} hideDisplay={hideDisplay}
    toggleDisplay={() => { }} />;
  render(component);

  await changeValue('Location name:', location);
  await changeValue('Longitude:', lon);
  await changeValue('Latitude:', lat);

  jest.spyOn(window, 'alert').mockImplementation(() => { });
  fetchMock.mockResponses(
    ['', { status: HttpStatus.ACCEPTED }],
    ['', { status: HttpStatus.OK }]);

  const button = await screen.findByText('Submit')
  fireEvent.click(button);
  await changeValue('Latitude:', '');


});

it('handles add location', async () => {
  const location = 'UCD';

  fetchMock.mockResponses(

    JSON.stringify({ value: false }),
    [
      JSON.stringify({ properties: { links: { harvest: '/harvest' } } }),
      { status: HttpStatus.OK }
    ],
    ['', { status: HttpStatus.ACCEPTED }],
    ['', { status: HttpStatus.OK }]);

  JSON.stringify({ properties: { links: { harvest: '/harvest' } } });
  jest.spyOn(window, 'alert').mockImplementation(() => { });
  const hideDisplay = jest.fn();


  const obtainData = () => {

    expect(alert).toBeCalledWith(location + ' was added.');

  };


  const component = <LocationForm display={true} containsRef={"http://localhost:8080/locationCollection/contains{?name}"} selfRef={"http://localhost:8080/locationCollection"}
    obtainData={obtainData} hideDisplay={hideDisplay}
    toggleDisplay={() => { }} />;
  render(component);

  await changeValue('Location name:', location);
  const button = await screen.findByText('Submit');
  fireEvent.click(button);
  await changeValue('Latitude:', '');
  expect(hideDisplay).toHaveBeenCalledTimes(1);

});

