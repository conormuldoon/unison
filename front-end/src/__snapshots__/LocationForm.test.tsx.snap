// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    id="pLeft"
  >
    <button
      data-testid="lf-button"
    >
      Hide
    </button>
    <form
      id="locForm"
    >
      <label>
        Location name: 
        <input
          name="name"
          required=""
          type="text"
          value=""
        />
      </label>
      <br />
      <br />
      <label>
        Longitude: 
        <input
          name="lon"
          required=""
          step="any"
          type="number"
          value=""
        />
      </label>
      <br />
      <br />
      <label>
        Latitude: 
        <input
          name="lat"
          required=""
          step="any"
          type="number"
          value=""
        />
      </label>
      <br />
      <br />
      <div>
        <input
          type="submit"
          value="Submit"
        />
      </div>
    </form>
  </div>
</div>
`;
